bitbots_vision/src/bitbots_vision/vision_modules/yolo_handler.py

```diff
@@ -1,111 +1,318 @@
+        Runs the neural network
+        """
+        # Check if cached
         if self.results is None:
+            # Run neural network
             self.results = self.net.detect(Image(self.image))
+            # Init lists
             class_ids = []
             confidences = []
             boxes = []
             self.ball_candidates = []
             self.goalpost_candidates = []
+            # Go through results
             for out in self.results:
+                # Get class id
                 class_id = out[0]
+                # Get confidence
                 confidence = out[1]
+                # Get candidate position and size
                 x, y, w, h = out[2]
                 x = x - int(w // 2)
                 y = y - int(h // 2)
-                c = Candidate(int(x), int(y), int(w), int(h))
-                c.rating = confidence
-                class_id = class_id
+                # Create candidate
+                c = Candidate(int(x), int(y), int(w), int(h), confidence)
+                # Append candidate to the right list depending on the class
                 if class_id == b"ball":
                     self.ball_candidates.append(c)
                 if class_id == b"goalpost":
                     self.goalpost_candidates.append(c)
 
-    def get_candidates(self):
-        self.predict()
-        return [self.ball_candidates, self.goalpost_candidates]
 
-class YoloBallDetector(CandidateFinder):
+class YoloHandlerOpenCV(YoloHandler):
+    """
+    Opencv library implementation of our yolo model
+    """
+    def __init__(self, config, model_path):
+        # Build paths
+        weightpath = os.path.join(model_path, "yolo_weights.weights")
+        configpath = os.path.join(model_path, "config.cfg")
+        # Set config
+        self.config = config
+        # Define classes
+        self.classes = ["ball", "goalpost"]
+        # Settings
+        self.nms_threshold = 0.4
+        self.confidence_threshold = 0.5
+        # Setup neural network
+        self.net = cv2.dnn.readNet(weightpath, configpath)
+        # Set default state to all cached values
+        self.image = None
+        self.blob = None
+        self.outs = None
+        self.goalpost_candidates = None
+        self.ball_candidates = None
+        self.results = None
 
-    def __init__(self, yolo):
+    @staticmethod
+    def get_output_layers(net):
+        """
+        Library stuff
+        """
+        layer_names = net.getLayerNames()
+
+        output_layers = [layer_names[i[0] - 1] for i in net.getUnconnectedOutLayers()]
+
+        return output_layers
+
+    def set_image(self, image):
+        """
+        Set a image for yolo. This also resets the caches.
+        :param image: current vision image
+        """
+        # Check if image has been processed
+        if np.array_equal(image, self.image):
+            return
+        # Set image
+        self.image = image
+        self.image = image
+        self.width = image.shape[1]
```

Comment:
Does this change during runtime? Could this be determined in the constructor?

Reply:
Not really but in some other parts of the vision its also done like that. The main advantage is, that this Vision is able to operate on different resolutions, without config changes. Runtime resolution changes are a nice benefit of it. It's also not difficult to calculate.

Reply:
I'm not even sure that it is calculated explicitly for this thing or whether it is necessary for numpy anyway. i don't see any reason to change this.